File: java/src/main/java/tech/qmates/openchat/AppFactory.java
package tech.qmates.openchat;

import tech.qmates.openchat.database.SQLitePostRepository;
import tech.qmates.openchat.database.SQLiteUserRepository;
import tech.qmates.openchat.domain.UTCClock;
import tech.qmates.openchat.domain.repository.PostRepository;
import tech.qmates.openchat.domain.repository.UserRepository;
import tech.qmates.openchat.domain.usecase.GetAllUserUseCase;
import tech.qmates.openchat.domain.usecase.GetTimelineUseCase;
import tech.qmates.openchat.domain.usecase.RegisterUserUseCase;
import tech.qmates.openchat.domain.usecase.SubmitPostUseCase;

import java.net.URISyntaxException;
import java.net.URL;
import java.time.ZoneId;
import java.time.ZonedDateTime;

public class AppFactory {

    private static UserRepository userRepositoryInstance;
    private static PostRepository postRepositoryInstance;

    public static GetTimelineUseCase buildGetTimelineUseCase() {
        return new GetTimelineUseCase(getPostRepository(), getUserRepository());
    }

    public static SubmitPostUseCase buildSubmitPostUseCase() {
        return new SubmitPostUseCase(getPostRepository(), getUserRepository(), getRealClock());
    }

    public static GetAllUserUseCase buildGetAllUserUseCase () {
        return new GetAllUserUseCase(AppFactory.getUserRepository());
    }

    public static RegisterUserUseCase buildRegisterUserUseCase() {
        return new RegisterUserUseCase(AppFactory.getUserRepository());
    }

    public static UserRepository getUserRepository() {
        if (userRepositoryInstance == null)
            userRepositoryInstance = new SQLiteUserRepository(getSqliteFilePath());

        return userRepositoryInstance;
    }

    public static PostRepository getPostRepository() {
        if (postRepositoryInstance == null)
            postRepositoryInstance = new SQLitePostRepository(getSqliteFilePath());

        return postRepositoryInstance;
    }

    public static void resetRepositories() {
        if (userRepositoryInstance != null)
            userRepositoryInstance.reset();
        if (postRepositoryInstance != null)
            postRepositoryInstance.reset();
    }

    public static UTCClock getRealClock() {
        return () -> ZonedDateTime.now(ZoneId.of("UTC"));
    }

    private static String getSqliteFilePath() {
        URL res = AppFactory.class.getClassLoader().getResource("production.db");
        if (res == null)
            throw new RuntimeException("Cannot find sqlite database file!");

        try {
            return res.toURI().getPath();
        } catch (URISyntaxException e) {
            throw new RuntimeException("Cannot load sqlite database file!", e);
        }
    }
}

File: typescript/src/AppFactory.ts
import SQLitePostRepository from "./database/SQLitePostRepository";
import SQLiteUserRepository from "./database/SQLiteUserRepository";
import PostRepository from "./domain/repositories/PostRepository";
import UserRepository from "./domain/repositories/UserRepository";
import GetAllUsersUseCase from "./domain/usecases/GetAllUsersUseCase";
import GetTimelineUseCase from "./domain/usecases/GetTimelineUseCase";
import RegisterUserUseCase from "./domain/usecases/RegisterUserUseCase";
import SubmitPostUseCase from "./domain/usecases/SubmitPostUseCase";

export default class AppFactory {
  private static userRepositoryInstance: UserRepository | null;
  private static postRepositoryInstance: PostRepository | null;

  static buildGetAllUsersUseCase() {
    return new GetAllUsersUseCase(AppFactory.getUserRepository())
  }

  static buildRegisterUserUseCase() {
    return new RegisterUserUseCase(AppFactory.getUserRepository())
  }

  static buildGetTimelineUseCase(): GetTimelineUseCase {
    return new GetTimelineUseCase(
      AppFactory.getPostRepository(),
      AppFactory.getUserRepository()
    )
  }

  static buildSubmitPostUseCase(): SubmitPostUseCase {
    return new SubmitPostUseCase(
      AppFactory.getPostRepository(),
      AppFactory.getUserRepository()
    )
  }

  static getUserRepository(): UserRepository {
    if (!this.userRepositoryInstance)
      this.userRepositoryInstance = new SQLiteUserRepository('src/database/production.db')

    return this.userRepositoryInstance
  }

  static getPostRepository(): PostRepository {
    if (!this.postRepositoryInstance)
      this.postRepositoryInstance = new SQLitePostRepository('src/database/production.db')

    return this.postRepositoryInstance
  }

  static resetRepositories(): void {
    this.userRepositoryInstance?.reset()
    this.postRepositoryInstance?.reset()
  }
}